crates/containerd-shim-wasmtime-v1/src/sandbox/manager.rs

```diff
@@ -24,33 +24,35 @@ use std::sync::Arc;
 use std::sync::RwLock;
 use std::thread;
 use ttrpc::context;
-use wasmtime::Engine;
 
-pub trait Sandbox: Task + Send + Sync {
-    type Instance: Instance;
+pub trait Sandbox<E>: Task + Send + Sync
+where
+    E: Send + Sync + Clone,
+{
+    type Instance: Instance<E = E>;
 
-    fn new(namespace: String, id: String, engine: Engine, publisher: RemotePublisher) -> Self;
+    fn new(namespace: String, id: String, engine: E, publisher: RemotePublisher) -> Self;
 }
 
-pub struct Service<T: Sandbox> {
+pub struct Service<E: Send + Sync + Clone, T: Sandbox<E>> {
     sandboxes: RwLock<HashMap<String, String>>,
-    engine: Engine,
+    engine: E,
     phantom: std::marker::PhantomData<T>,
 }
 
-impl<T: Sandbox> Service<T> {
-    pub fn new(engine: Engine) -> Self {
-        Service::<T> {
+impl<E: Send + Sync + Clone, T: Sandbox<E>> Service<E, T> {
+    pub fn new(engine: E) -> Self {
+        Service::<E, T> {
             sandboxes: RwLock::new(HashMap::new()),
             engine: engine,
             phantom: std::marker::PhantomData,
         }
     }
 }
 
-impl<T> Manager for Service<T>
+impl<E: Send + Sync + Clone, T> Manager for Service<E, T>
 where
-    T: Sandbox<Instance = WasiInstance> + 'static,
+    T: Sandbox<E, Instance = WasiInstance> + 'static,
```

Comment:
Can `E`'s type constraints be moved into the `where` clause here?

Reply:
Resolved